### ISSUE TITLE
Support User Defined Functions on Athena Dialect

### ISSUE THREAD

**Issue by sbrandtb** [tobymao/sqlglot#3087]
It looks like sqlglot is not able to parse [AWS Athena's user defined functions syntax](https://docs.aws.amazon.com/athena/latest/ug/querying-udf.html):

```py
from sqlglot import parse
from sqlglot.dialects import Trino

parse("""
    USING EXTERNAL FUNCTION some_function(input VARBINARY)
        RETURNS VARCHAR
        LAMBDA 'some-name'
    SELECT
        some_function(1)
""", dialect=Trino)
```

Exception:

```
sqlglot.errors.ParseError: Invalid expression / Unexpected token. Line 2, Col: 9.
  
    USING EXTERNAL FUNCTION some_function(input VARBINARY)
        RETURNS VARCHAR
        LAMBDA 'some-name'
```

We are using `Trino` dialect since sqlglot does not have a dedicated one for Athena, as far as I understand, but Athena is based off Trino, so this dialect works otherwise perfectly for our codebase :slightly_smiling_face: 

Am I missing something? Does it need a dedicated dialect for Athena?

**Comment by georgesittas** [comment]
Hey @sbrandtb, thanks for the report.

AWS Athena is not supported right now, so this behavior is expected as it's not valid _Trino_ syntax, IIUC. Closing this as not planned for now since we don't plan to add Athena in the near future.

**Comment by sbrandtb** [comment]
@georgesittas Given that Athena is basically Trino, what speaks against deriving a new Athena dialect, starting with UDF syntax? I am willing to help.

**Comment by georgesittas** [comment]
We actually just discussed this internally and decided to implement the `Athena` dialect and have this be parsed into a `Command` for now. Should be a good enough compromise to start with. Appreciate your willingness to help here.

**Comment by georgesittas** [comment]
So should have a fix for this particular issues soon. Feel free to contribute improvements to Athena if you find more discrepancies in the future.

**Comment by sbrandtb** [comment]
@georgesittas I really appreciate your effort and fast response, but as far as I understand, the solution implemented in #3089 returns a `Command`, which does not include the Expressions/Abstract Syntax Tree that SELECT statements normally parse into. 

I see how my expectation ("no error raised") was very misleading, sorry for that :sweat_smile: 

I took some time and implemented a parser for the UDF declaration statement. Good for us, we do not require the information from the declaration, so I simply parse it, forget about it and carry on with SELECT parsing:

```py

class Athena(Trino):
    class Parser(Trino.Parser):
        def _parse_udf_declaration(self) -> None:
            if not self._match_text_seq('EXTERNAL', 'FUNCTION'):
                self.raise_error('Expecting EXTERNAL FUNCTION')

            if not self._parse_id_var():
                self.raise_error('Expecting function name')

            self._match_l_paren()

            while not self._match(TokenType.R_PAREN):
                if not self._parse_type():
                    self.raise_error('Expecting data type')

            if not self._match_texts('RETURNS'):
                self.raise_error('Expecting RETURNS')

            if not self._parse_type():
                self.raise_error('Expecting data type')

            if not self._match_texts('LAMBDA'):
                self.raise_error('Expecting LAMBDA')

            if not self._match(TokenType.STRING):
                self.raise_error('Expected lambda function name string')

        def _parse_using(self) -> Expression | None:
            """
            Parses Athena's UDF declaration, but throws away results since we
            are not interested in them so far.

            https://docs.aws.amazon.com/athena/latest/ug/querying-udf.html
            """
            self._parse_udf_declaration()

            while self._match(TokenType.COMMA):
                self._parse_udf_declaration()

            return self._parse_select()

        STATEMENT_PARSERS = {
            **Trino.Parser.STATEMENT_PARSERS,
            TokenType.USING: _parse_using,
        }
```

My first try in sqlglot. What do you think?

**Comment by georgesittas** [comment]
> I took some time and implemented a parser for the UDF declaration statement. Good for us, we do not require the information from the declaration, so I simply parse it, forget about it and carry on with SELECT parsing:

We shouldn't drop any part of the query, because then we're breaking it. This is why I converted the whole thing into a `Command` - because then it's preserved.

I'd rather we add a new `udfs` arg in `Select` to store the inline functions and implement the parsing and generation correctly so that they're roundtripped. This may be helpful if we later decide to add support for Trino's [inline routines](https://trino.io/docs/current/routines/introduction.html#inline-routines), too.
